### PR TITLE
TDI-36685: improve the error information

### PR DESCRIPTION
**Pull request by wwang-talend**
https://jira.talendforge.org/browse/TDI-36685

### CODE REVIEW

File: components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java [hunk 1]
```diff
@@ -213,4 +225,39 @@ private OutputStream writeToServer(String requestMethod, String request) throws
 
         return metaData.getColumns();
     }
+
+    private String extractResponseInformationAndConsumeResponse(HttpResponse response) throws IllegalStateException, IOException {
+        InputStream is = response.getEntity().getContent();
+        return extractResponseInformationAndConsumeResponse(is);
```
Comment (sgandon) [inline components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java]: I would suggest a try/finally block here to close the input stream that you created.
It is usually better to have the code that closes the stream at the same level as the one opening it.

File: components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java [hunk 2]
```diff
@@ -213,4 +225,39 @@ private OutputStream writeToServer(String requestMethod, String request) throws
...
+    }
+
+    private String extractResponseInformationAndConsumeResponse(HttpURLConnection connection) throws IOException {
+        InputStream is = null;
+        try {
+            is = connection.getInputStream();
+        } catch (IOException e) {
+            is = connection.getErrorStream();
+        }
+        return extractResponseInformationAndConsumeResponse(is);
```
Comment (sgandon) [inline components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java]: same comment here.

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetinput/TDataSetInputProperties.java [hunk 3]
```diff
@@ -78,8 +80,19 @@ public TDataSetInputProperties(String name) {
...
         Form form = getForm(Form.MAIN);
-        form.addRow(dataSetName);
+        form.addRow(Widget.widget(this.dataSetName).setWidgetType("widget.type.dataset.selection"));
         form.addRow(Widget.widget(fetchSchema).setWidgetType(Widget.BUTTON_WIDGET_TYPE));
+
+        Form advancedForm = new Form(this, Form.ADVANCED);
+        advancedForm.addRow(dataSetId);
+    }
+
+    public static final String ADD_QUOTES = "ADD_QUOTES";
```
Comment (sgandon) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetinput/TDataSetInputProperties.java]: What is the reason for the ADD_QUOTES ?
In the component there should not be any knowledge of the Studio and therefore the idea that the Studio uses quotes or not for a value is not of the concern of the Component Developer.

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java [hunk 4]
```diff
@@ -132,11 +180,62 @@ public RuntimeProperties getRuntimeProperties() {
...
+            runtimeProperties.setDataSetName(dataSetNameForCreateMode.getStringValue());
         } else {
-            runtimeProperties.setDataSetName(dataSetId.getStringValue());
+            runtimeProperties.setDataSetId(dataSetId.getStringValue());
         }
         runtimeProperties.setLimit(limit.getStringValue());
         return runtimeProperties;
     }
+
+    public ValidationResult afterFetchSchema() {
```
Comment (sgandon) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: I don't see any unit test for this method, please provide one.

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetinput/TDataSetInputProperties.java (hunk 3 above)
Comment (wwang-talend) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetinput/TDataSetInputProperties.java]: dataSetName is a special field without double quote, and can't set it to hidden. So when generate the code, a compile issue appear. Avoid this issue by this code.

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java (hunk 4 above)
Comment (wwang-talend) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetoutput/TDataSetOutputProperties.java]: have added one.

File: components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java (hunk 1 above)
Comment (wwang-talend) [inline components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java]: thanks for your advice, have fixed it.

File: components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetinput/TDataSetInputProperties.java (hunk 3 above)
Comment (sgandon) [inline components-dataprep/src/main/java/org/talend/components/dataprep/tdatasetinput/TDataSetInputProperties.java]: what does it make it special, all String field do not have quotes anyway from the Component developer point of view.

File: components-dataprep/src/test/java/org/talend/components/dataprep/tdatasetinput/AfterFetchSchemaTest.java [hunk 5]
```diff
@@ -35,21 +36,35 @@
...
+    // TODO should remove the duplicated code for afterFetchSchema
+    @Test
+    public void testAfterFetchSchemaForOutput() {
+        TDataSetOutputProperties properties = (TDataSetOutputProperties) componentService.getComponentDefinition("tDatasetOutput")
+                .createProperties();
+        properties.url.setValue("http://localhost:" + serverPort);
+        properties.login.setValue("vincent@dataprep.com");
+        properties.pass.setValue("vincent");
+        properties.dataSetName.setValue("mydataset");
+        properties.dataSetId.setValue("db119c7d-33fd-46f5-9bdc-1e8cf54d4d1e");
```
Comment (sgandon) [inline components-dataprep/src/test/java/org/talend/components/dataprep/tdatasetinput/AfterFetchSchemaTest.java]: Please provide also test cases for the different possible errors

File: components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java [hunk 6]
```diff
@@ -227,18 +227,31 @@ private OutputStream writeToServer(String requestMethod, String request) throws
     }
 
     private String extractResponseInformationAndConsumeResponse(HttpResponse response) throws IllegalStateException, IOException {
-        InputStream is = response.getEntity().getContent();
-        return extractResponseInformationAndConsumeResponse(is);
+        InputStream is = null;
```
Comment (sgandon) [inline components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java]: I would suggest this.
```
InputStream is = response.getEntity().getContent();
 try {
       return extractResponseInformationAndConsumeResponse(is);
} finally {
       is.close();
}
```

File: components-dataprep/src/test/java/org/talend/components/dataprep/tdatasetinput/AfterFetchSchemaTest.java (hunk 5 above)
Comment (wwang-talend) [inline components-dataprep/src/test/java/org/talend/components/dataprep/tdatasetinput/AfterFetchSchemaTest.java]: get it, will do it, but before creating the cases, i need to move afterFetchSchema to the parent class from TDataSetOutputProperties and TDataSetInputProperties.

File: components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java (hunk 6 above)
Comment (wwang-talend) [inline components-dataprep/src/main/java/org/talend/components/dataprep/connection/DataPrepConnectionHandler.java]: agree, to do it.

**Pull request merged**